### PR TITLE
Fix wrong classify behavior if the string is already classified

### PR DESCRIPTION
**Pull request by choallin**
See issue #208

### CODE REVIEW

File: lib/hanami/utils/string.rb [hunk 1]
```diff
@@ -151,6 +152,10 @@ def classify
         self.class.new words.zip(delimiters).join
       end
 
+      def classified?
+        split(NAMESPACE_SEPARATOR).reduce(false) {|memo, str| str.match(/\A[A-Z][a-zA-Z_]*/).to_s.length == str&.length }
```
Comment (davydovanton) [inline lib/hanami/utils/string.rb]: I think it's so complicated method and we need to simplify it.

That's why I suggest:
1. Set regexp to constant
2. Instead `#reduce` you can use [`#all?`](https://apidock.com/ruby/Enumerable/all%3F) method
3. I think you can use `/\w+/` instead `/[a-zA-Z_]/`

File: lib/hanami/utils/string.rb [hunk 2]
```diff
@@ -141,6 +141,7 @@ def capitalize
       #   string = Hanami::Utils::String.new 'hanami_utils'
       #   string.classify # => 'HanamiUtils'
       def classify
+        return self.class.new to_s if classified?
```
Comment (davydovanton) [inline lib/hanami/utils/string.rb]: I think we can return `self` here because we don't mutate state 🤔

File: test/string_test.rb [hunk 3]
```diff
@@ -85,6 +87,20 @@
...
+    it 'correctly recognizes a classified String' do
+      Hanami::Utils::String.new('hanami').classified?.must_equal(false)
+      Hanami::Utils::String.new('Hanami_router').classified?.must_equal(true)
+      Hanami::Utils::String.new('hanami-router').classified?.must_equal(false)
+      Hanami::Utils::String.new('Hanami/router').classified?.must_equal(false)
+      Hanami::Utils::String.new('Hanami/router').classified?.must_equal(false)
+      Hanami::Utils::String.new('Hanami/Router').classified?.must_equal(false)
+      Hanami::Utils::String.new('HanamiRouter').classified?.must_equal(true)
+      Hanami::Utils::String.new('AwesomeProject::Namespace').classified?.must_equal(true)
+      Hanami::Utils::String.new('AwesomeProject::SubNamespace').classified?.must_equal(true)
```
Comment (davydovanton) [inline test/string_test.rb]: Really strange logic for `Hanami_router` and `Hanami/Router`, but it just my opinion 😃

File: lib/hanami/utils/string.rb (hunk 2 above)
Comment (choallin) [inline lib/hanami/utils/string.rb]: You are right, self should be fine here!

File: test/string_test.rb (hunk 3 above)
Comment (choallin) [inline test/string_test.rb]: Since ```Hanami_router``` is a valid class name it should be true

**Comment by choallin** [comment]
I will also comment the ```classified?``` method

File: lib/hanami/utils/string.rb (hunk 1 above)
Comment (mereghost) [inline lib/hanami/utils/string.rb]: Also, classes can legally have numbers on it's names, so `\w` would be more suitable as @davydovanton said. =)

**Comment by mereghost** [comment]
Oh, @choallin could you fix the rubocop warning so the CI build will pass? =)

**Comment by artofhuman** [comment]
Should we write tests with `RSpec` instead `Minitest`. In master we alread run all tests by `RSpec`
https://github.com/hanami/utils/blob/master/Rakefile#L24

**Comment by davydovanton** [comment]
awesome catch @jodosha @artofhuman 👏

File: test/string_test.rb [hunk 4]
```diff
@@ -85,6 +87,20 @@
     end
   end
 
+  describe '#classified' do
```
Comment (jodosha) [inline test/string_test.rb]: @choallin You introduced unit tests for `#classified?`, but not for the original issue, which was related to `#classify`.

File: lib/hanami/utils/string.rb [hunk 5]
```diff
@@ -141,6 +147,7 @@ def capitalize
       #   string = Hanami::Utils::String.new 'hanami_utils'
       #   string.classify # => 'HanamiUtils'
       def classify
+        return self if classified?
```
Comment (jodosha) [inline lib/hanami/utils/string.rb]: @choallin This looks like a workaround that doesn't fix the real problem in the following lines. Would you like to check why it doesn't work properly without adding this `#classfied?` guard?

---

More on the guard. Guards are useful if they are inexpensive. This is not the case: `#classified?` requires a lot of computations.

File: lib/hanami/utils/string.rb [hunk 6]
```diff
@@ -151,6 +158,24 @@ def classify
...
+      #
+      # @example
+      #   require 'hanami/utils/string'
+      #
+      #   string = Hanami::Utils::String.new 'hanami_utils'
+      #   string.classified? # => false
+      #
+      #   string = Hanami::Utils::String.new 'HanamiUtils'
+      #   string.classified? # => true
+      def classified?
```
Comment (jodosha) [inline lib/hanami/utils/string.rb]: If we decide to go with `#classified?` this should be a private method and `@api private`.

File: test/string_test.rb (hunk 4 above)
Comment (choallin) [inline test/string_test.rb]: @jodosha I added the lines 73 & 74 to test the original issue. Should I make a new test case for that?

File: lib/hanami/utils/string.rb (hunk 5 above)
Comment (choallin) [inline lib/hanami/utils/string.rb]: @jodosha I will take a look at the regex/method again if I can make it without it.

File: test/string_test.rb (hunk 4 above)
Comment (jodosha) [inline test/string_test.rb]: @choallin Nope, sorry. I missed that lines. That tests are enough.

**Comment by TiteiKo** [comment]
I think one of the current tests is wrong, making everything way more complex...

```ruby
expect(Hanami::Utils::String.new('hanami-router').classify).to eq('Hanami::Router')
```

As 'hanami-router' underscored equals 'hanami_router' and not 'hanami/router', and as `#classify` relies on `#underscore`, then  it should produce 'HanamiRouter' instead of 'Hanami::Router'.

If we consider that this test is incorrect, then fixing the `#classify` method becomes way simpler:

```diff
diff --git a/lib/hanami/utils/string.rb b/lib/hanami/utils/string.rb
index 5ee64fa..4da25ab 100644
--- a/lib/hanami/utils/string.rb
+++ b/lib/hanami/utils/string.rb
@@ -142,7 +142,7 @@ module Hanami
       #   string.classify # => 'HanamiUtils'
       def classify
         words = underscore.split(CLASSIFY_WORD_SEPARATOR).map!(&:capitalize)
-        delimiters = scan(CLASSIFY_WORD_SEPARATOR)
+        delimiters = underscore.scan(CLASSIFY_WORD_SEPARATOR)

         delimiters.map! do |delimiter|
           delimiter == CLASSIFY_SEPARATOR ? EMPTY_STRING : NAMESPACE_SEPARATOR
diff --git a/spec/unit/hanami/utils/string_spec.rb b/spec/unit/hanami/utils/string_spec.rb
index 8138b85..3d86001 100644
--- a/spec/unit/hanami/utils/string_spec.rb
+++ b/spec/unit/hanami/utils/string_spec.rb
@@ -64,16 +64,18 @@ RSpec.describe Hanami::Utils::String do
     it 'returns a classified string' do
       expect(Hanami::Utils::String.new('hanami').classify).to eq('Hanami')
       expect(Hanami::Utils::String.new('hanami_router').classify).to eq('HanamiRouter')
-      expect(Hanami::Utils::String.new('hanami-router').classify).to eq('Hanami::Router')
+      expect(Hanami::Utils::String.new('hanami-router').classify).to eq('HanamiRouter')
       expect(Hanami::Utils::String.new('hanami/router').classify).to eq('Hanami::Router')
       expect(Hanami::Utils::String.new('hanami::router').classify).to eq('Hanami::Router')
       expect(Hanami::Utils::String.new('hanami::router/base_object').classify).to eq('Hanami::Router::BaseObject')
+      expect(Hanami::Utils::String.new('AwesomeProject::Namespace').classify).to eq('AwesomeProject::Namespace')
+      expect(Hanami::Utils::String.new('AwesomeProject::SubNamespace').classify).to eq('AwesomeProject::SubNamespace')
     end

     it 'returns a classified string from symbol' do
       expect(Hanami::Utils::String.new(:hanami).classify).to eq('Hanami')
       expect(Hanami::Utils::String.new(:hanami_router).classify).to eq('HanamiRouter')
-      expect(Hanami::Utils::String.new(:'hanami-router').classify).to eq('Hanami::Router')
+      expect(Hanami::Utils::String.new(:'hanami-router').classify).to eq('HanamiRouter')
       expect(Hanami::Utils::String.new(:'hanami/router').classify).to eq('Hanami::Router')
       expect(Hanami::Utils::String.new(:'hanami::router').classify).to eq('Hanami::Router')
     end
```

**Pull request merged**